### DISCUSSION TITLE
Able to delete multiple records using batchWrite even when where clause is not satisfied

### DISCUSSION THREAD

**Discussion by hemanth1226** [sensedeep/dynamodb-onetable#66] [Q&A] [answered]
Hi, I am trying to delete multiple records conditionally using batchWrite. But the records are getting deleted even when condition is not satisfied.
Please find the code below and let me know how we can fix it. When I try the same with just deleteItem, its working fine.

I have a record in dynamo DB with 
id: '01A7XV8933CP2XMYA40X3ACX4Q',
userId: 'test123'

when a user with id 'random123' try to delete the record, as the condition is not satisfied, i am expecting to see error returned. `deleteItem` is working as expected but `batchWrite` is deleting record.

batch Code:
``` javascript()
public async deleteItems(ids: string[], userId: string): Promise<{}> {
    const batch = {};
    ids.forEach(async (id) => {
      await this.model.deleteItem(
        { id: id },
        {
          batch,
          many: true,
          where: '${userId} = {userId}',
        },
      );
    });
    return await this.table.batchWrite(batch);
  }
```

Individual record deletion:
``` javascript()
public async deleteItems(ids: string[], userId: string): Promise<void> {
    return this.model.deleteItem(
        { id: ids[0] },
        {
          many: true,
          where: '${userId} = {userId}',
        },
      );
  }
```

**Comment by mobsense** [comment]
What are your table keys?  Can you post the schema. When you delete, you need to provide the full key for the item to delete.

Also, include Table params.logger == true and put {log: true} in your API params so we can see the actual command sent to DynamoDB.

You only use "many"true" when your delete call matches multiple items. It is a safety feature if your delete call actually matches more than one item.

**Comment by hemanth1226** [comment]
Please find the schema and logs below. I also removed `many:true` which is not required. 
``` javascript()
schema = {
    indexes: {
      primary: { hash: 'pk', sort: 'sk' },
      gsi1: { hash: 'gsi1pk', sort: 'gsi1sk', project: 'all', follow: true },
      gsi2: { hash: 'gsi2pk', sort: 'gsi2sk', follow: true },
      gsi3: { hash: 'gsi3pk', sort: 'gsi3sk', follow: true },
    },
    models: {
      Mailbox: {
        pk: { type: String, value: 'mail#${id}' },
        sk: { type: String, value: 'mail#${id}' },
        read: { type: Boolean },
        subject: { type: String },
        desc: { type: String },
        body: { type: String },
        userId: { type: String, required: true },
        id: { type: String, required: true, uuid: 'ulid' },
        gsi1pk: {
          type: String,
          value: 'user#${userId}',
        },
        gsi1sk: {
          type: String,
          value: 'mail#${id}',
        },
        gsi2pk: {
          type: String,
          value: 'user#${userId}#unread',
        },
        gsi2sk: {
          type: String,
          value: 'mail#${id}',
        },
        _type: { type: String, filter: false, value: 'Mailbox' },
      },
    },
  };
```

``` javascript()
table = new Table({
    client: client,
    name: 'table-name',
    delimiter: '#',
    schema: schema,
    logger: true
  });
```
``` javascript()
  model = this.table.getModel('Mailbox');
```

Please find the log below:
``` javascript()
"batch": {
        "RequestItems": {
            "table-name": [
                {
                    "DeleteRequest": {
                        "Key": {
                            "pk": {
                                "S": "mail#01A7XV8933CP2XMYA40X3ACX4Q"
                            },
                            "sk": {
                                "S": "mail#01A7XV8933CP2XMYA40X3ACX4Q"
                            }
                        }
                    }
                },
                {
                    "DeleteRequest": {
                        "Key": {
                            "pk": {
                                "S": "mail#01A7XV8933CP2XMYA40X3ACX4Z"
                            },
                            "sk": {
                                "S": "mail#01A7XV8933CP2XMYA40X3ACX4Z"
                            }
                        }
                    }
                }
            ]
        }
    }
```

**Comment by mobsense** [comment, marked as answer]
Thanks. Just checked the Bible, with Batch Delete, you can't have filter conditions.

We'll add a warning to the delete API if you are using batch with where.

Generally, if you have a query (including delete), you want to always use a key rather than a non-key attribute. In your case, gsi1 would locate by user id without doing a scan.

So this should do it for you.

    return this.model.deleteItem({userId}, {index: 'gsi1'})

**Reply by hemanth1226** [reply]
Hi I tried deleteItem by passing userId and mail Id as shown below. I have multiple records with that userId and only need to delete the record that matches the mail id. If this works, i will work on batch deletion. 
``` javascript()
return this.model.deleteItem(
      {
       userId,
       id: ids[0]
      },
      {
        index: 'gsi1',
      },
    );
```

This is what i am seeing in logs:
``` javascript()
error Dynamo exception in "delete" on "Mailbox" {
    "err": {
        "__type": "com.amazon.coral.validate#ValidationException",
        "name": "ValidationException",
        "$fault": "client",
        "$metadata": {
            "httpStatusCode": 400,
            "requestId": "K1TQPR6AJ1FM15L9N9KRS520BJVV4KQNSO5AEMVJF66Q9ASUAAJG",
            "attempts": 1,
            "totalRetryDelay": 0
        }
    },
    "trace": {
        "cmd": {
            "TableName": "table-name",
            "Key": {}
        },
        "op": "delete",
        "properties": {
            "userId": "test123",
            "id": "01A7XV8933CP2XMYA40X3ACX4Q"
        },
        "params": {
            "index": "gsi1",
            "checked": true,
            "fallback": true
        },
        "err": {
            "__type": "com.amazon.coral.validate#ValidationException",
            "name": "ValidationException",
            "$fault": "client",
            "$metadata": {
                "httpStatusCode": 400,
                "requestId": "K1TQPR6AJ1FM15L9N9KRS520BJVV4KQNSO5AEMVJF66Q9ASUAAJG",
                "attempts": 1,
                "totalRetryDelay": 0
            }
        }
    }
}
```

Can you please help me understand what i am doing wrong here?

**Reply by mobsense** [reply]
You are using deleteItem instead of remove. 

Use Model.remove. The definitions in model.d.ts for the *Item methods should not be there and we'll remove them as they are private internal methods only.

Normally with DynamoDB you cannot do delete via a GSI. However, OneTable will do a query on the GSI to get the primary index keys and then reissue the delete on the primary index. So the following works with OneTable.

```
await this.model.remove({userId}, {index: 'gsi1', log: true})
```

However, if you already have the ID which makes up the primary hash key, you can do it directly on the primary index:

```
await this.model.remove({id})
```

The Table class does have a public low-level API but it is meant for operating on DynamoDB when you don't have an entity with a "type". i.e. legacy data.

**Reply by mobsense** [reply]
One last comment back to the OP about batches.

If you have the IDs which are the primary key, then you can use a batch with just that key:

```
let batch = {}
myModel.remove({id}, {batch})
await table.batchWrite(batch)
```

NOTE: the delete does NOT await and the batchWrite DOES await.  Your OP was waiting on the deleteItem and was using the deleteItem and not model.remove
